### DISCUSSION TITLE
New Tagged/discriminated enums and new match operator for pattern matching

### DISCUSSION THREAD

**Discussion by khalyomede** [aardappel/lobster#242] [Ideas]
Hi, just learned about Lobster on Youtube. Big fan of the simplicity and syntax, congratulations!

The proposal below assumes you answer yes to the question "Can you/do you want to implement it". Feel free to totally ditch it if it's not the case!

## Tagged enum - typed constructor

We could define enums cases to accept constructors:

```julia
import steam

enum Result:
  ok: fn (string)
  error: fn (string)

def get_file_content(path: string) -> Result:
  content = steam_read_file(path)

  if content is nil:
    return error("Could not open file")

  return ok(content)

content = match file_get_content("does-not-exist.yml")
  ok (value: string): value
  error (message: string): message

print(content) # "Could not open file"

content = match file_get_content("README")
  ok (value: string): value
  error (message: string): message

print(content) # "# Hello world"
```

## Tagged enum - with generic parameter

```julia
import steam

enum Result<T, E>:
  ok: fn (T)
  error: fn (E)

struct Exception:
  message: string
  stack = []

def get_file_content(path: string) -> Result<string, Exception>:
  content = steam_read_file(path)

  if content is nil:
    return error( Exception { message: "Could not open file" } )

  return ok(content)

content = match file_get_content("does-not-exist.yml")
  ok (value: string): value
  error (exception: Exception): exception.message

print(content) # "Could not open file"

content = match file_get_content("README")
  ok (value): value
  error (exception: Exception): exception.message

print(content) # "# Hello world"
```

## Why

I think lobster did an an amazing job at implementing type safety in a simple manner. I always wondered through different statically analyzed languages (F#, Haskell, V, Zig, ...) and every time I tried to put my head in those, I always stumble upon a "what the freck" moment about either the syntax, or the poor tagged union capability/support (to be completely fair, F# was the closest to my definition of a good and simple language).

I think lobster can definitely differenciate itself by implement **true** tagged union and pattern matching ability. Languages have always wanted to draw a line between FP and OOP, and Lobster can join them both like a good old Lord of the Ring movie.

More seriously, in my opinion having Tagged union and pattern matching would increase the robustness of the language because this way of describing a program is way much safer since the programmer that use your module not only handle the "happy path" but also all the possible "exception".

## End note

This proposal will have no sense if you tell me the compiler is able to statically warn if a code that could throw is not properly try-catched. 

The big show stopper about languages implementing throw mecanism is mostly the inability to be able to detect unhandled possible exceptions thrown by a function. If Lobster is able to detect this, then this proposal would have a lower interest for me.

**Comment by aardappel** [comment]
Yes, I have thoughts about adding tagged unions at some point. Though note:

* Pattern matching is an orthogonal concern from tagged unions, and can equally well be performed on objects. See for example here about adding pattern matching to objects: https://github.com/aardappel/lobster/issues/113
  * People sometimes like tagged unions because dispatching code for them can be kept next to eachother, as opposed to spread over classes. Lobster can already do that with non-inline methods declared right next to eachother.
* Tagged unions are in some cases all about performance (in Rust, not most FPLs), since they can be in-line allocated rather than dynamically allocated. If max performance is not your concern, then existing objects are largely equivalent.
  * This performance hinges on all cases of the union having very similar amounts of fields. For example if you make a Rust enum where one case has 1 field and the other has 5 fields, and the 1 field case is most frequently used, this is wildly inefficient.
* Tagged unions can have different semantics because they are by value rather than by reference, allowing you to guarantee instances are never shared. If that is not a concern for you, or you actually may want sharing, objects again work just as well.
* People generally like them a lot in Rust because Rust (and FPLs) don't have any good traditional inheritance mechanisms, but Lobster does.

That is to say, people tend to have a lot of excitement for tagged unions but are not actually aware why they want them (over object with inheritance). Personally, my biggest interest from the above would be in performance, I am guessing for you it is pattern matching? What if Lobster had pattern matching on objects, would you still want them?

Anyway, if I'd add them to Lobster, they'd likely piggy-back on `struct` rather than trying to overload `enum` with new meaning.

```
struct Result
struct Ok : Result
struct Error : Result
    err:string
```

Or maybe we could invent a more convenient syntax to all declare that in 1, but that is immaterial to the concept.

Then, an instance of `Result` would be the union of sizes of its "subclasses", and have a hidden type field, all passed by value like current structs already are.

If you look in `TODO.TXT` in the repo, you'd find these notes on tagged unions (which I call ADTs, from the FPL world):

```
- NOTE: now that we have inline structs, ADTs should be based on them, since being same-sized
  and being inline go so wonderfully together!
  - This would make types like Maybe/Optional etc super cheap.
  - don't even need to invent some separate enum/adt feature and use pattern matching for this,
    structs that participate in an inheritance hierarchy can simply have a typeinfo prefixed,
    ONLY when they are in a context of the superclass! That is even more efficient than enums
    in Rust, which presumably always carry their tag. In fact it is more similar to "fat" trait
    pointers in Rust, except without the pointers. A tagged struct can even shrink to not need
    the padding to bring all these structs to the same size.. and all this is easy because they're
    always copied.
- Rust's same-size ADTs with value semantics allow interesting copy-over use cases that are not
  possible in even C++, where there is no need to update pointers to an object.
  Maybe we can have a declaration in Lobster that forces a certain set of derived classes to be
  padded to the same size, with a special overwrite operator. The type of the variable would need
  to be of the superclass, and the superclass itself should maybe not be instantiable, or also be
  the same size.
  Basically you'd declare the superclass as everything below here is same size.
  What are examples of algorithms that could be simpler this way?
  - game objects changing state
  - AST node being changed
  - Type variables! This one is hard to do without.
  - Values in an interpreter
  Though most of these are easy to do with existing code.. it is more to avoid allocation and
  allow these values to be inline? Without inline values, you first have to construct on the heap
  before overwrite. Still useful to be possible for cases like type variables.
- Features to make it easier to use structs ADT style as opposed to inheritance
  - ADT is similar to a base class with no members and a fixed set of sub-classes.
    - So make that use case into a simpler declaration syntax.
  - extend switch to do pattern matching.
    - failing that, a switch that can use types as case labels would be a good start,
      and simple to do given that we have type indices!
```

**Reply by khalyomede** [reply]
Thank you for giving your honest point of view about it. If Lobster had pattern matching over object this would be sufficient enough for my use case (and as you suggested I would represent my ADTs using structs). In fact, I mentioned generic tagged unions, but personally this would not be of a big interest for me to be honest. Again structs would be fine for me!

As long as Lobster correct me when I forget to check an error branch in my code, and I do not have errors at runtime, I'm happy and open to anything. End goal: to sleep at night 😂

Happy to know this was already on your todo list and you considered them by design. I really like your honesty thank you again!

**Comment by aardappel** [comment]
Oh and as for your note on exceptions: Lobster does not have language-native exceptions. It does however have them implemented in terms of non-local returns in `exceptions.lobster`, and if you do a `throw` thats not inside a `try` you will indeed get an error.

**Reply by khalyomede** [reply]
Very cool, then since there is some kind of safety on this area, I guess my need for tagged enums/pattern matching is less important. While this could be a good idea to implement it, it has the drawback of overloading/complexifying the language. Maybe other folks will find other useful use-cases but for me that will do it.

Thank you for your instructive and insightful view on this subject. I learned on the princess and I felt heard. That means a lot to me. Hope the community sticks to this way of behaving, keep up the good work!

**Comment by aardappel** [comment]
For now, we have exhaustive switch on types, which goes pretty far towards this functionality, is simple and fast!

Example: https://github.com/aardappel/lobster/blob/b74f00c97cf9f43093e24f73dc8b1ccef063d46f/tests/typeswitch.lobster#L6-L77
Docs: https://github.com/aardappel/lobster/blob/b74f00c97cf9f43093e24f73dc8b1ccef063d46f/docs/source/language_reference.md?plain=1#L727-L750
